gradio/routes.py

```diff
@@ -311,17 +310,19 @@ def login(form_data: OAuth2PasswordRequestForm = Depends()):
         def main(request: fastapi.Request, user: str = Depends(get_current_user)):
             mimetypes.add_type("application/javascript", ".js")
             blocks = app.get_blocks()
-            root_path = route_utils.get_root_url(request)
+            root = route_utils.get_root_url(request)
             if app.auth is None or user is not None:
-                config = copy.deepcopy(app.get_blocks().config)
-                config["root"] = root_path
-                config = add_root_url(config, root_path)
+                config = app.get_blocks().config
+                previous_root = config.get("root", None)
+                if previous_root is None or previous_root != root:
```

Comment:
The main change in this PR being that we no longer need to traverse the config if the `root` has not change